### PR TITLE
Automatically fast-forward and reattach head

### PR DESCRIPTION
**Pull request by xendk**
As discussed in #158 .

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Please gate automatic fast-forward with an option, I like getting prompted so I can look at what commits are new. (I don't care if the option is on by default)

File: straight.el [hunk 1]
```diff
@@ -1598,56 +1598,37 @@ confirmation, so this function should only be run after
...
                            "git" "merge-base" "--is-ancestor"
                            branch ref-name))
              (ref-behind-p (straight--check-call
                             "git" "merge-base" "--is-ancestor"
                             ref-name branch)))
-         (when (and ref ref-behind-p)
+         (when ref-ahead-p
+           ;; Local is behind, catch up.
+           (straight--get-call "git" "reset" "--hard" ref-name)
            (cl-return-from straight-vc-git--validate-head t))
```
Comment (raxod502) [inline straight.el]: I don't think we should be returning non-nil here. Usually there will be more work to do with cleanup after a reset, no?

File: straight.el [hunk 2]
```diff
@@ -1598,56 +1598,37 @@ confirmation, so this function should only be run after
...
        (let ((ref-ahead-p (straight--check-call
                            "git" "merge-base" "--is-ancestor"
                            branch ref-name))
              (ref-behind-p (straight--check-call
                             "git" "merge-base" "--is-ancestor"
                             ref-name branch)))
-         (when (and ref ref-behind-p)
+         (when ref-ahead-p
+           ;; Local is behind, catch up.
+           (straight--get-call "git" "reset" "--hard" ref-name)
```
Comment (raxod502) [inline straight.el]: This should only happen automatically in the case that `ref` is non-nil, I think.

File: straight.el [hunk 3]
```diff
@@ -1598,56 +1598,37 @@ confirmation, so this function should only be run after
...
            (cl-return-from straight-vc-git--validate-head t))
+         (when ref-behind-p
+           (if ref (cl-return-from straight-vc-git--validate-head t)
+             (straight--get-call "git" "checkout" branch)))
```
Comment (raxod502) [inline straight.el]: This definitely should never happen automatically, as it's not a fast-forward.

File: straight.el [hunk 4]
```diff
@@ -1598,56 +1598,37 @@ confirmation, so this function should only be run after
...
-                                branch quoted-ref-name)
-                   ,(lambda ()
-                      (straight--get-call
-                       "git" "reset" "--hard" ref-name)))))
-            ,@(when (and ref-behind-p (null ref))
-                `(("f" ,(format "Fast-forward HEAD to branch %S" branch)
-                   ,(lambda ()
-                      (straight--get-call
-                       "git" "checkout" branch)))))
-            ,@(unless (or ref-ahead-p ref-behind-p)
```
Comment (raxod502) [inline straight.el]: These changes should be conditional on a new option, as discussed above.

File: straight.el (hunk 1 above)
Comment (xendk) [reply straight.el]: There is? `git reset --hard` reverts the index and the worktree, it's always been my get-out-of-jail card when I've messed up, so I don't know what cleanup you're referring to?

File: straight.el (hunk 2 above)
Comment (xendk) [reply straight.el]: Well, if `ref` is nil, `ref-name` is `"HEAD"`, so it basically checks if the branch is even with itself (detached heads and other branches are dealt with earlier), so the reset should be a no-op in most cases (dirty worktrees should have been dealt with in `straight-vc-git--validate-worktree`).

So yeah, you're right, it should depend on `ref`, at least for efficiency.

File: straight.el (hunk 3 above)
Comment (xendk) [reply straight.el]: I... Have no idea what I was thinking.

File: straight.el (hunk 1 above)
Comment (raxod502) [reply straight.el]: Take a look at this example:

```
example-git-repo % git init
Initialized empty Git repository in /private/tmp/example-git-repo/.git/
example-git-repo[master] % git commit --allow-empty -m "Initial commit"
[master (root-commit) 352e524] Initial commit
example-git-repo[master] % echo 'ignored-file' > .gitignore
example-git-repo[master*] % touch ignored-file
example-git-repo[master*] % git add .gitignore
example-git-repo[master*] % git commit -m "Add .gitignore"
[master e5420a0] Add .gitignore
 1 file changed, 1 insertion(+)
 create mode 100644 .gitignore
example-git-repo[master] % git status
On branch master
nothing to commit, working tree clean
example-git-repo[master] % git reset --hard @~
HEAD is now at 352e524 Initial commit
example-git-repo[master*] % git status
On branch master
Untracked files:
        ignored-file

nothing added to commit but untracked files present
```

File: README.md [hunk 5]
```diff
@@ -1666,6 +1666,10 @@ You can customize the following user options:
   operations will offer to re-set your remote URLs from HTTPS to SSH
   or vice versa, depending on the value of
   `straight-vc-git-default-protocol`. This is nil by default.
+* `straight-vc-git-force-protocol`: if this is non-nil, pulling will
```
Comment (raxod502) [inline README.md]: Do you mean `straight-vc-git-auto-fast-forward`?

**Comment by xendk** [comment]
OK, take two, now less ambitious and with an option.

I've renamed all the validate functions, as they all potentially change the repository.

File: straight.el (hunk 1 above)
Comment (xendk) [reply straight.el]: So we should do a `git clean -df` afterwards?

File: README.md (hunk 5 above)
Comment (xendk) [reply README.md]: Doh... Rewriting history...

File: straight.el (hunk 1 above)
Comment (raxod502) [reply straight.el]: No, just return nil. If you look at how I've structured the validation functions, each one returns nil if some previously established state (such as a clean worktree) might have been destroyed. Then, I chain them together by running the sequence of validation functions until all of them return non-nil. In that case, the desired validation has been effected.

This should probably be documented somewhere, oops.

Comment (xendk) [reply straight.el]: Oooh, right.

Fixed.

File: straight.el [hunk 6]
```diff
@@ -1433,7 +1441,7 @@ Do not suppress unexpected errors."
 
 ;;;;;; Validation functions
 
-(cl-defun straight-vc-git--validate-remote (local-repo remote desired-url)
+(cl-defun straight-vc-git--ensure-remote (local-repo remote desired-url)
   "Validate that LOCAL-REPO has REMOTE set to DESIRED-URL or equivalent.
```
Comment (raxod502) [inline straight.el]: If we are renaming `validate` to `ensure`, should we also change the docstring to reflect that terminology?

File: straight.el [hunk 7]
```diff
@@ -1596,13 +1604,10 @@ confirmation, so this function should only be run after
...
        (error "Branch %S does not exist" ref))
       ((and (null ref) (string= branch cur-branch))
-       (cl-return-from straight-vc-git--validate-head t))
+       (cl-return-from straight-vc-git--ensure-head t))
       ((and (null ref) head-detached-p)
-       (straight-vc-git--popup
-         (format "In repository %S, HEAD is even with branch %S, but detached."
-                 local-repo branch)
-         ("a" (format "Attach HEAD to branch %S" branch)
-          (straight--get-call "git" "checkout" branch))))
```
Comment (raxod502) [inline straight.el]: Can we make this only happen automatically if `straight-vc-git-auto-fast-forward` is non-nil? It's a nontrivial change to the state of the repository on disk.

**Comment by xendk** [comment]
Good comments, I've fixed.

**Comment by raxod502** [comment]
Looks like there is a long line somewhere, according to the build failure.

File: straight.el [hunk 8]
```diff
@@ -1284,6 +1284,15 @@ one specified in `straight-vc-git-default-protocol'."
   :type 'boolean
   :group 'straight)
 
+(defcustom straight-vc-git-auto-fast-forward t
+  "Whether to quietly fast-forward when pulling packages.
+This suppresses popups for trivial remote changes (i.e. the
+current HEAD is an ancestor to the remote HEAD).
+Also re-attaches detatced heads quietly when non-nil.
```
Comment (raxod502) [inline straight.el]: s/detatced/detached

**Comment by xendk** [comment]
Aaah, the joys of late night pushes... Fixed.

**Comment by raxod502** [comment]
Thanks! This is great.

**Pull request merged**